src/test/ringdelaybuffer_test.cpp

```diff
@@ -0,0 +1,182 @@
+            numSamples);
+
+    AssertIdenticalBufferEquals(output.span(),
+            mixxx::spanutil::spanFromPtrLen(thirdExpectedResult, numSamples));
+}
+
+TEST_F(RingDelayBufferTest, ReadWriteDelayTest) {
+    const SINT numSamples = 4;
+    const SINT firstDelaySize = 2;
+    const SINT secondDelaySize = 4;
+    const CSAMPLE inputBuffer[] = {-100.0, 100.0, -99.0, 99.0};
+    const CSAMPLE firstExpectedResult[] = {-50.0, 50.0, -99.0, 99.0};
+    const CSAMPLE secondExpectedResult[] = {0.0, 0.0, -50.0, 50.0};
+    const CSAMPLE thirdExpectedResult[] = {-50.0, 50.0, -99.0, 99.0};
+
+    mixxx::SampleBuffer output(numSamples);
+
+    // Read without delay.
+    EXPECT_EQ(m_pRingDelayBuffer->write(
+                      mixxx::spanutil::spanFromPtrLen(inputBuffer, numSamples)),
+            numSamples);
+    EXPECT_EQ(m_pRingDelayBuffer->read(
+                      mixxx::spanutil::spanFromPtrLen(output.data(), numSamples), 0),
+            numSamples);
+
+    AssertIdenticalBufferEquals(output.span(),
+            mixxx::spanutil::spanFromPtrLen(firstExpectedResult, numSamples));
+
+    // Read with delay.
+    EXPECT_EQ(m_pRingDelayBuffer->read(
+                      mixxx::spanutil::spanFromPtrLen(output.data(), numSamples), firstDelaySize),
+            numSamples);
+
+    AssertIdenticalBufferEquals(output.span(),
+            mixxx::spanutil::spanFromPtrLen(secondExpectedResult, numSamples));
+
+    // Fill the second half of the delay buffer with the first input data.
+    EXPECT_EQ(m_pRingDelayBuffer->write(
+                      mixxx::spanutil::spanFromPtrLen(inputBuffer, numSamples)),
+            numSamples);
+
+    // Read with delay (not circle around).
+    EXPECT_EQ(m_pRingDelayBuffer->read(
+                      mixxx::spanutil::spanFromPtrLen(output.data(), numSamples), secondDelaySize),
+            numSamples);
+
+    AssertIdenticalBufferEquals(output.span(),
+            mixxx::spanutil::spanFromPtrLen(thirdExpectedResult, numSamples));
+}
+
+static void BM_WriteReadWholeBufferNoDelay(benchmark::State& state) {
+    const SINT ringDelayBufferSize = static_cast<SINT>(state.range(0));
+    const SINT numSamples = ringDelayBufferSize / 2;
+
+    RingDelayBuffer m_ringDelayBuffer(ringDelayBufferSize);
+
+    mixxx::SampleBuffer input(numSamples);
+    mixxx::SampleBuffer output(numSamples);
+
+    SampleUtil::fill(input.data(), 0.0f, numSamples);
+
+    for (auto _ : state) {
+        state.PauseTiming();
+        m_ringDelayBuffer.clear();
+        state.ResumeTiming();
+
+        m_ringDelayBuffer.write(mixxx::spanutil::spanFromPtrLen(input.data(), numSamples));
+        m_ringDelayBuffer.read(mixxx::spanutil::spanFromPtrLen(output.data(), numSamples), 0);
+        m_ringDelayBuffer.write(mixxx::spanutil::spanFromPtrLen(input.data(), numSamples));
+        m_ringDelayBuffer.read(mixxx::spanutil::spanFromPtrLen(output.data(), numSamples), 0);
```

Comment:
Didn't we already merge the PR containing `SampleBuffer::span`?
It would cut down on boilerplate code considerably in all tests. 
```suggestion
        m_ringDelayBuffer.write(input.span());
        m_ringDelayBuffer.read(output.span(), 0);
        m_ringDelayBuffer.write(input.span());
        m_ringDelayBuffer.read(output.span(), 0);
```